File: JULY-2025/16-07-2025/p3201.erl
-spec maximum_length(Nums :: [integer()]) -> integer().
maximum_length(Nums) ->
    %% init C01 to 1, since we count from 2nd element flip (first is always the same)
    find_parity_flip(Nums, hd(Nums) rem 2, 1, 0, 0) .

%% truth table setup for parity and 
find_parity_flip([A | Nums], Parity, C01, C0, C1) -> 
    if
        Parity == 0 andalso A rem 2 == 0 -> find_parity_flip(Nums, A rem 2, C01  , C0+1, C1);
        Parity == 0 andalso A rem 2 == 1 -> find_parity_flip(Nums, A rem 2, C01+1, C0  , C1+1);
        Parity == 1 andalso A rem 2 == 0 -> find_parity_flip(Nums, A rem 2, C01+1, C0+1, C1);
        Parity == 1 andalso A rem 2 == 1 -> find_parity_flip(Nums, A rem 2, C01  , C0  , C1+1);
        true -> err
    end;
find_parity_flip([], Parity, C01, C0, C1) -> lists:max([C01, C0, C1]).

File: JULY-2025/16-07-2025/p3201.rs
impl Solution {
    pub fn maximum_length(nums: Vec<i32>) -> i32 {
        let a = nums
            .iter()
            .fold([0, 0, 0, nums[0] & 1], |[odd, even, flip, last], n| {
                let curr = n & 1;
                [
                    odd + curr,
                    even + (1 - curr),
                    flip + (last ^ curr),
                    last ^ (last ^ curr),
                ]
            });
        a[0].max(a[1].max(a[2] + 1))
    }
}